fluentinterface/src/main/java/com/iluwatar/fluentinterface/fluentiterable/lazy/DecoratingIterator.java

```diff
@@ -0,0 +1,53 @@
+package com.iluwatar.fluentinterface.fluentiterable.lazy;
+
+import java.util.Iterator;
+
+/**
+ * This class is used to realize LazyFluentIterables. It decorates
+ * a given iterator.
+ * @param <TYPE>
+ */
+public abstract class DecoratingIterator<TYPE> implements Iterator<TYPE> {
+
+    protected final Iterator<TYPE> fromIterator;
+
+    private TYPE next = null;
+
+    /**
+     * Creates an iterator that decorates the given iterator.
+     * @param fromIterator
+     */
+    public DecoratingIterator(Iterator<TYPE> fromIterator) {
+        this.fromIterator = fromIterator;
+    }
+
+    /**
+     * Precomputes and caches the next element of the iteration.
+     * @return true if a next element is available
+     */
+    @Override
+    public final boolean hasNext() {
+        next = computeNext();
```

Comment:
Guava `AbstractIterable` has a `state` property which changes to `READY` when `computeNext()` returns a element. And then on consecutive calls to `hasNext()`, if `state` is `READY` then it returns `true` and does not compute next element.
Like this:
```java
public boolean hasNext() {
		switch (state) {
		case READY:
			return true;
		case DONE:
			return false;
		default:
		}
		return tryComputeNext();
	}
```